Get htrace.sh to work in macOS

Niek opened the pull request.

adamdavis40208 commented:
Just adding my two cents: installing from your branch still leaves me with:

```
$htrace.sh -d http://nmap.org -s -h
not found in PATH: geoiplookup mixed-content-scan
```

(and I don't know enough about composer to know why it adds to path or if it's just a "my box" kind of issue)

Full terminal output:


```
$htrace.sh git:(master) npm install -g observatory-cli
/usr/local/bin/observatory -> /usr/local/lib/node_modules/observatory-cli/index.js
+ observatory-cli@0.7.1
updated 1 package in 4.197s


   ╭───────────────────────────────────────────────────────────────╮
   │                                                               │
   │       New minor version of npm available! 6.5.0 → 6.8.0       │
   │   Changelog: https://github.com/npm/cli/releases/tag/v6.8.0   │
   │               Run npm install -g npm to update!               │
   │                                                               │
   ╰───────────────────────────────────────────────────────────────╯

$htrace.sh git:(master) composer global require bramus/mixed-content-scan
Changed current directory to /Users/adavis/.composer
Using version ^2.9 for bramus/mixed-content-scan
./composer.json has been updated
Loading composer repositories with package information
Updating dependencies (including require-dev)
Nothing to install or update
Generating autoload files
```

Niek commented:
I see I forgot to add the install instructions for ``geoiplookup`` to the README, fixed now. I also added a fix for people who don't have the composer global dir in their PATH like you - that should fix the missing mixed-content-scan binary.

adamdavis40208 commented:
I still had to edit my path to add composure (after rerunning setup), but other than that it worked great!

![image](https://user-images.githubusercontent.com/19850646/53769550-13825780-3e91-11e9-8cbb-823a094bdced.png)

trimstray commented:
Thanks for this, I'll test it on my high sierra virtualbox instance.

berlincount commented on bin/htrace.sh:
```diff
@@ -39,6 +39,18 @@
+  [ ! -z "$(composer global config bin-dir --absolute 2>/dev/null)" ] && PATH=$(composer global config bin-dir --absolute 2>/dev/null):$PATH
+
+# shellcheck disable=SC2003,SC2046
+elif [[ "$(expr substr $(uname -s) 1 5)" == "Linux" ]] ; then
```
why is there a substr necessary? I only ever see `uname -s` return "Linux" on Linux.

also, if there's an `else` that always returns `true` anyways .. what`s the point?

I kind of dislike disabling `shellcheck` strong .. (ditto in `setup.sh`)

trimstray replied:
1) I used a better solution for this
2) For the future
3) Yes, you're right, I'll correct it

berlincount commented:
```
$ sudo ./dependencies.sh install
readlink: illegal option -- f
usage: readlink [-n] [file ...]


  Before start please read all communication on the screen!


  os_name: macos
    os_id: macos

  Please enter (Y) to continue: y
$ htrace.sh -u http://www.snafu.de --hide-src-ip

     htrace.sh v1.1.1  (openssl : not tested)


    URI: http://www.snafu.de ; Method: GET

    req  time_total   time_connect    local_socket           via              remote_socket         geo   proto   ver   code     next_hop
    ---  ----------   ------------    ------------           ---              -------------         ---   -----   ---   ----     --------
 •   1   0.047299     0.047299        xxx.xxx.xxx.xxx:55858  xxx.xxx.xxx.xxx  213.73.83.45:80             http    1.1   301      https://www.snafu.de/
 •   2   1.436525     1.389226        xxx.xxx.xxx.xxx:55860  xxx.xxx.xxx.xxx  213.73.83.45:443            https   1.1   200
```

seems to behave as expected :)

berlincount commented:
```
$ htrace.sh -u http://nmap.org -s -h --hide-src-ip --observatory

     htrace.sh v1.1.1  (openssl : not tested)


    URI: http://nmap.org ; Method: GET

    req  time_total   time_connect    local_socket           via              remote_socket         geo   proto   ver   code     next_hop
    ---  ----------   ------------    ------------           ---              -------------         ---   -----   ---   ----     --------
 •   1   0.359878     0.359878        xxx.xxx.xxx.xxx:55915  xxx.xxx.xxx.xxx  45.33.49.119:80             http    1.1   301      https://nmap.org/
         ssl: off

         › HTTP/1.1 301 Moved Permanently
         › Date: Tue, 05 Mar 2019 16:11:08 GMT
         › Server: Apache/2.4.6 (CentOS)
         › Location: https://nmap.org/
         › Content-Length: 298
         › Keep-Alive: timeout=5, max=100
         › Connection: Keep-Alive
         › Content-Type: text/html; charset=iso-8859-1

    req  time_total   time_connect    local_socket           via              remote_socket         geo   proto   ver   code     next_hop
    ---  ----------   ------------    ------------           ---              -------------         ---   -----   ---   ----     --------
 •   2   1.271200     0.911322        xxx.xxx.xxx.xxx:55918  xxx.xxx.xxx.xxx  45.33.49.119:443            https   1.1   200
         ssl: on, version(TLSv1.2), cipher(ECDHE-RSA-AES128-GCM-SHA256)
         public-key(2048 bit), signature(sha256WithRSAEncryption)
         date: Mar 16 00:00:00 2018 GMT / Mar 15 23:59:59 2020 GMT
         issuer: <empty>
         owner: <empty>
         cn: <empty>
         san: nmap.org www.nmap.org
         sni: match
         validity: match
         verification:

         › HTTP/1.1 200 OK
         › Date: Tue, 05 Mar 2019 16:11:12 GMT
         › Server: Apache/2.4.6 (CentOS)
         › Strict-Transport-Security: max-age=31536000; preload
         › Accept-Ranges: bytes
         › Keep-Alive: timeout=5, max=100
         › Connection: Keep-Alive
         › Transfer-Encoding: chunked
         › Content-Type: text/html; charset=utf-8

    Mozilla Observatory (https://observatory.mozilla.org)

     ▸ https://nmap.org

         › HTTP Observatory Report: nmap.org
         › Score Rule                           Description
         ›   -50 subresource-integrity          Subresource Integrity (SRI) not implemented, and external scripts are loaded over HTTP or use protocol-relative URLs via src="//...".
         ›   -25 content-security-policy        Content Security Policy (CSP) header not implemented.
         ›   -20 x-frame-options                X-Frame-Options (XFO) header not implemented.
         ›   -10 x-xss-protection               X-XSS-Protection header not implemented.
         ›    -5 x-content-type-options         X-Content-Type-Options header not implemented.
         ›     0 contribute                     Contribute.json isn't required on websites that don't belong to Mozilla.
         ›     0 cookies                        No cookies detected.
         ›     0 cross-origin-resource-sharing  Content is not visible via cross-origin resource sharing (CORS) files or headers.
         ›     0 public-key-pinning             HTTP Public Key Pinning (HPKP) header not implemented.
         ›     0 redirection                    Initial redirection is to HTTPS on same host, final destination is HTTPS.
         ›     0 referrer-policy                Referrer-Policy header not implemented.
         ›     0 strict-transport-security      HTTP Strict Transport Security (HSTS) header set to a minimum of six months (15768000).
         › Score: 0
         › Grade: F
         › Full Report Url: https://observatory.mozilla.org/analyze.html?host=nmap.org
```

also ok (Mojave) .. `dependencies.sh` still broken, though

trimstray commented:
@Niek @berlincount

```
$ sudo ./dependencies.sh install
readlink: illegal option -- f
usage: readlink [-n] [file ...]
```

for this i should use `gnu-getopt` (to fix) and:

```
Please enter (Y) to continue: y
```

`Y` not equal `y`

trimstray commented:
One more things:

```
htrace.sh v1.1.1  (openssl : not tested)
```

What version do you use?

berlincount commented:
... of what? :thinking:

trimstray commented:
`openssl` ;) this mechanism incorrectly checks the `openssl` version.

berlincount commented:
oh, hah.

```
$ openssl version -a
LibreSSL 2.6.5
built on: date not available
platform: information not available
options:  bn(64,64) rc4(ptr,int) des(idx,cisc,16,int) blowfish(idx)
compiler: information not available
OPENSSLDIR: "/private/etc/ssl"
```

uh.

macOS Mojave 10.14.3

Niek commented:
This was fixed in https://github.com/trimstray/htrace.sh/pull/59
